src/reducers/collapseContext.ts

```diff
@@ -1,15 +1,26 @@
-import { parentOf, pathToContext, reducerFlow, unroot } from '../util'
+import { head, parentOf, pathToContext, reducerFlow, unroot } from '../util'
 import { State } from '../util/initialState'
-import { getChildren } from '../selectors'
+import { getAllChildren, simplifyPath } from '../selectors'
 import { archiveThought, existingThoughtMove, setCursor } from '../reducers'
+import _ from 'lodash'
+import existingThoughtDelete from './existingThoughtDelete'
+import { Path } from '../types'
+
+interface Options {
+  deleteCursor?: boolean,
+  at?: Path | null,
+}
 
 /** Collapses the active thought. */
-const collapseContext = (state: State) => {
+const collapseContext = (state: State, { deleteCursor, at }: Options) => {
   const { cursor } = state
 
   if (!cursor) return state
 
-  const children = getChildren(state, pathToContext(cursor))
+  const simpleCursor = simplifyPath(state, cursor)
```

Comment:
`at` is supposed to completely replace the `cursor`, e.g.: https://github.com/cybersemics/em/blob/cd80687634aff5c34b7f9e0b03fa0a16a77f2e75/src/reducers/newThought.ts#L94 It allows any `Path` to be collapsed. An interface that operates partly on the cursor and partly on an arbitrary `Path` would be confusing. 

At the risk of saying things you already know, let me share a little more of my thought process. I understand it may not apply completely if the dual `Path` approach is needed here, but hopefully you can find something useful in it anyway :).

I ask myself, does the interface accurately reflect the purpose of this function? Am I adding an option  only as an implementation detail, where it doesn't relate directly to the purpose? There are some cases where this is necessary, in which case comments are essential. Otherwise I want to be very careful about how I am designing the internal API.

In this case we have a straightforward reducer that archives a context and moves its children to its parent. It makes sense that you may want to delete instead of archive, but not always, so we make that an optional argument. It also makes sense that we want to be able to collapse any `Path`, not just the `cursor`, so we have a way to set the `Path` and default to the `cursor`. `at` is a variable name that reflects where the operation will be executed. We could also opt for a more explicit and less "natural language" approach and call it `path`, although the `at` convention is used elsewhere so I opted for that. If two paths are needed (the site of the operation, and where to move the cursor to), then we should rename `at` to something that reflects that its more narrow purpose. Note that we should also consider the possibility of removing the `setCursor` completely from the reducer, and setting the cursor in the action creator instead. However, the `cursor` movement is considered to be atomic in most cases; if the operation is performed on the `cursor`, the `cursor` should be updated. Otherwise it's possible to leave the `cursor` in an invalid state (such as a `Path` that has been moved or deleted). So this leads us to automatic `cursor` movement as the default, with some options to override if necessary (like in `importText`).

Reply:
@raineorshine Thanks for the detailed thought process. I now properly understand the use of `at`. I was trying to achieve different thing using `at`. I was using it to override default `cursor` setting logic of `collapseContext` and directly set given `path` as the cursor passed using `at`. Back of the mind I was questioning myself that this doesn't reflect the purpose of the function. I was very wrong.

Reply:
No worries! I see now that `at` was clever but properly not the best variable name to use :).